### PR TITLE
repair CI for Win

### LINKED ISSUE #2347: Error when importing metrics on Windows without DDP support
## 🐛 Bug

When loading the new metrics the following AttributeError is raised:  

`AttributeError: module 'torch.distributed' has no attribute 'ReduceOp'`. 

The problem is the use of `torch.distributed.ReduceOp` in the type hints and some functions of `pytorch_lightning.metrics.converters.py`.

A similar (identical?) issue was discussed on the [NVIDIA/apex github](https://github.com/NVIDIA/apex/issues/429). The issues seems to be due to the fact that Windows does not support certain distributed features, and as a result the `ReduceOp` class is completely missing in the Windows binaries. Their approach was to guard those imports with `if torch.distributed.is_available()`.

### To Reproduce

```python
import pytorch_lightning.metrics
```

### Expected behavior

Metrics should only raise an error if I try to use DDP on a machine without the necessary support.

### Environment

* CUDA:
	- GPU:
		- GeForce GTX 1050 Ti with Max-Q Design
	- available:         True
	- version:           10.2
* Packages:
	- numpy:             1.18.4
	- pyTorch_debug:     False
	- pyTorch_version:   1.5.1
	- pytorch-lightning: 0.8.1
	- tensorboard:       2.2.1
	- tqdm:              4.46.0
* System:
	- OS:                Windows
	- architecture:
		- 64bit
		- WindowsPE
	- processor:         Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
	- python:            3.7.4
	- version:           10.0.17763

### LINKED ISSUE #2357: accuracy metric dosen't support windows
## 🐛 Bug

Pytorch Metric.Accuracy uses `ReduceOp` from 'torch.distribution' but torch.distributrion doesn't support `windows`
 - https://github.com/pytorch/pytorch/blob/cf8a9b50cacb1702f5855859c657a5358976437b/torch/distributed/__init__.py#L10 : `torch.distributed is available on Linux and MacOS.` 

### To Reproduce

Use Metric.Accuracy in Windows environment

#### Code sample

- I use code sample from `https://github.com/PyTorchLightning/pytorch-lightning/issues/2355`

### Expected behavior

add check OS in `metric.accuracy` and use condition for import different module 
``` 
  try:
    return platform.linux_distribution()
  except:
    return "N/A"
```
or warning to windows user, they can't use `metric.accuracy`

### Environment

```
* CUDA:
        - GPU:
                - GeForce RTX 2080 Ti
                - GeForce GTX 1080 Ti
        - available:         True
        - version:           10.1
* Packages:
        - numpy:             1.18.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.5.1
        - pytorch-lightning: 0.8.1
        - tensorboard:       2.2.1
        - tqdm:              4.46.0
* System:
        - OS:                Windows
        - architecture:
                - 64bit
                - WindowsPE
        - processor:         Intel64 Family 6 Model 158 Stepping 10, GenuineIntel
        - python:            3.6.10
        - version:           10.0.18362
```
### Additional context
```
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    from pytorch_lightning.metrics.functional import accuracy
  File "C:\Users\dcho\Anaconda3\envs\torch_py36\lib\site-packages\pytorch_lightning\metrics\__init__.py", line 1, in <module>
    from pytorch_lightning.metrics.converters import numpy_metric, tensor_metric
  File "C:\Users\dcho\Anaconda3\envs\torch_py36\lib\site-packages\pytorch_lightning\metrics\converters.py", line 220, in <module>
    reduce_op: Optional[torch.distributed.ReduceOp] = None,
AttributeError: module 'torch.distributed' has no attribute 'ReduceOp'
```

Always thanks for developing & maintaining the cool framework

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

it seems like Win CI was falling but the checks were passed
https://github.com/PyTorchLightning/pytorch-lightning/pull/2356/checks?check_run_id=805809911

The issue was reported in https://github.com/nedbat/coveragepy/issues/1003
Fixes #2347
Fixes #2357

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
failing for Windows:
```
C:\hostedtoolcache\windows\Python\3.6.8\x64\lib\site-packages\py\_path\local.py:704: in pyimport
    __import__(modname)
pytorch_lightning\metrics\__init__.py:1: in <module>
    from pytorch_lightning.metrics.converters import numpy_metric, tensor_metric
pytorch_lightning\metrics\converters.py:220: in <module>
    reduce_op: Optional[torch.distributed.ReduceOp] = None,
E   AttributeError: module 'torch.distributed' has no attribute 'ReduceOp'
```

**Comment by justusschock** [comment]
@Borda the distributed part is fixed. only the horovod part still persists

**Comment by awaelchli** [comment]
horovod is not supported on windows, so we need to somehow exclude it from installing, not sure how.

**Comment by Borda** [comment]
@tgaddair we may have some problem with Horovod on Win, is it supported?
Otherwise, we may skip it for now :] @justusschock ^^

**Comment by tgaddair** [comment]
Hey @Borda, Horovod is not supported on Windows.  Thanks for checking!

File: pytorch_lightning/metrics/converters.py [hunk 1]
```diff
@@ -10,8 +10,15 @@
...
-
 from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities import rank_zero_warn
+
+try:
+    from torch.distributed import ReduceOp
+except ImportError:
+    class ReduceOp:
+        SUM = None
+    rank_zero_warn('Unsupported `ReduceOp` for distributed computing.')
```
Comment (Borda) [inline pytorch_lightning/metrics/converters.py:21]:
```suggestion

    rank_zero_warn('Unsupported `ReduceOp` for distributed computing.')
```

**Comment by Borda** [comment]
@tgaddair this is very strange as it is in try/except
```
tests\models\data\horovod\train_default_model.py:25: in <module>
    import horovod.torch as hvd
E   ModuleNotFoundError: No module named 'horovod
```

**Comment by Borda** [comment]
@awaelchli @justusschock @PyTorchLightning/core-contributors any idea why it fails even during the test collection before running any test? just for Windows... :/

**Comment by Borda** [comment]
fixed the failing but for some reason, the Win CI is super slow...

**Comment by awaelchli** [comment]
The tests became slow for me since the ddp mode was switched from spawn. I'm pretty sure that's it.

**Comment by Borda** [comment]
> The tests became slow for me since the ddp mode was switched from spawn. I'm pretty sure that's it.

so shall we test by default `ddp_spawn`?

**Comment by williamFalcon** [comment]
sure that would work. but we need at least one ddp test.

**Comment by awaelchli** [comment]
try num_workers=0 in tests. it will speed up on windows. but probably on linux and mac it will be slower then.

**Comment by Borda** [comment]
> sure that would work. but we need at least one ddp test.

it runs all tests for all ddp as before just replace the default one, so all functionality is tested

**Comment by justusschock** [comment]
@awaelchli IIRC this comes from the way, windows Handles process creation. On linux and Mac you usually create a process by forking it which then comes with a repopulated process context. On windows a new process is created with empty context always and the overhead comes from the time that is necessary to create the context state, defined by the parent process

**Comment by Borda** [comment]
@awaelchli your proposal with nb workers helped so MUCH, not it takes 6min and before 60min
anyway these tests shall be fixed in #2375 
```
FAILED tests/trainer/test_dataloaders.py::test_val_dataloader_not_implemented_error[1.0]
FAILED tests/trainer/test_dataloaders.py::test_train_dataloader_not_implemented_error[50]
FAILED tests/trainer/test_dataloaders.py::test_train_dataloader_not_implemented_error[1.0]
FAILED tests/trainer/test_dataloaders.py::test_val_dataloader_not_implemented_error_failed
FAILED tests/trainer/test_dataloaders.py::test_test_dataloader_not_implemented_error_failed
```

**Comment by awaelchli** [comment]
@yukw777 In #1667 you added a tmpdir_server fixture to load checkpoints from urls, but now it looks like the url lib has a problems with windows. have been looking at this for 2h plus but can't figure it out. Any ideas what could be the problem?
(not saying it's your fault, the tests for windows were not running when you made the PR)

**Comment by Borda** [comment]
@awaelchli mind this help with paths #2294?

**Comment by awaelchli** [comment]
as I said in my previous comment, I am trying to fix this for hours now but no luck so far.

**Comment by yukw777** [comment]
@awaelchli, sorry for the late response! I’ve been traveling for the past few days and just saw this. It seems like you guys figured it out though. What was the issue with Windows?

File: tests/conftest.py [hunk 2]
```diff
@@ -48,7 +48,7 @@ def translate_path(self, path):
         class ThreadingHTTPServer(ThreadingMixIn, HTTPServer):
             daemon_threads = True
 
-    with ThreadingHTTPServer(('', 0), Handler) as server:
+    with ThreadingHTTPServer(('localhost', 0), Handler) as server:
```
Comment (yukw777) [inline tests/conftest.py:51]: Ah man!!

Comment (awaelchli) [reply tests/conftest.py:51]: hey no problem. 
yep that was it, Windows didn't like the 0.0.0.0 address :)

**Comment by Borda** [comment]
There were a couple of them as all Win tests were ignored for a few weeks or months...

**Pull request merged**